### PR TITLE
[SKU Scanner] Handle barcode formats

### LINKED ISSUE #9757: [SKU Scanner] Test for different barcode standards
From review: https://github.com/woocommerce/woocommerce-ios/pull/9736#issuecomment-1551880585

> it's probably worth a short investigation of UPC-A barcodes. I've found that those I generate sometimes get returned by the scanner with a leading zero number I used

If we need to target specific barcode, while this is internal, we can access through the `VNDetectBarcodesRequest.symbologies` enum:

```
 23 elements
  VNBarcodeSymbologyAztec
  VNBarcodeSymbologyCodabar
  VNBarcodeSymbologyCode128
  VNBarcodeSymbologyCode39
  VNBarcodeSymbologyCode39Checksum
  VNBarcodeSymbologyCode39FullASCII
  VNBarcodeSymbologyCode39FullASCIIChecksum
  VNBarcodeSymbologyCode93
  VNBarcodeSymbologyCode93i
  VNBarcodeSymbologyDataMatrix
  VNBarcodeSymbologyEAN13
  VNBarcodeSymbologyEAN8
  VNBarcodeSymbologyGS1DataBar
  VNBarcodeSymbologyGS1DataBarExpanded
  VNBarcodeSymbologyGS1DataBarLimited
  VNBarcodeSymbologyI2of5
  VNBarcodeSymbologyI2of5Checksum
  VNBarcodeSymbologyITF14
  VNBarcodeSymbologyMicroPDF417
  VNBarcodeSymbologyMicroQR
  VNBarcodeSymbologyPDF417
  VNBarcodeSymbologyQR
  VNBarcodeSymbologyUPCE
```

The one linked in the issue is UPC-A, which would be scanned under EAN13 symbology.

### PR DESCRIPTION
**Pull request by toupper**
Closes: #9757

## Description

With this PR we handle different barcode formats and their peculiarities so we can adapt the SKU search to its requirements. In order to achieve that we refactor the barcode scanning logic so we can also retrieve the barcode format, and we encapsulate the barcode-SKU search into its own class `BarcodeSKUScannerProductFinder`, where we add the formats handling specific logic. As this struct uses `async/await` for its API, we adapt its clients to that.

## Testing instructions

Please test these cases:

1. Open the app
2. Tap on barcode scan button on the Order list, or tap to create a new order and find the barcode scanning button

### Happy Path:
Barcode matches your product SKU

### Your SKU doesn't have a check digit:
For UPC-A (EAN-13) and UPC-E, test the barcode scanning when your product SKU doesn't have the scanned check digit (last one).

### UPC-A:
As Apple converts 12-digit UPC-A barcodes into EAN-13 by adding a 0 at the beginning, please test when your SKU is in UPC-A format, both when your product SKU has the check digit and when not.

---
As this PR includes some logic refactoring in both scanning entry points, please test that they work properly.

You can use [this](https://barcode.tec-it.com/en/UPCA?data=72527273070) excellent resource to generate barcodes. Please check [here](pdfdoF-2Zq-p2) for a deeper reference about this issue.

## Screenshots

N/A

---
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Products/SKU Scanner/BarcodeScannerViewController.swift [hunk 1]
```diff
@@ -2,6 +2,13 @@ import AVFoundation
 import UIKit
 import Vision
 
+typealias BarcodeSymbology = VNBarcodeSymbology
+
+struct ScannedBarcode: Equatable, Hashable {
+    let payloadStringValue: String
+    let symbology: BarcodeSymbology
+}
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Products/SKU Scanner/BarcodeScannerViewController.swift:10]: 💯

**Review by iamgabrielma — APPROVED** [review]
Tests well! 

I found an issue when attempting to scan an UPC-E via the Order Creation view, it didn't work on the first time and didn't show any notice either. The same happened with an entirely different QR for a product variation in the same screen (Order Creation). 

This was resolved once I scanned these via the Order List scanner, and then attempted to scan again via Order Creation, so most likely is not related to this PR.

In the Order List view it always worked when supposed to work, and returned a failure notice when was supposed to fail, so the issue seems to be narrowed to the Order Creation one.

If helps, you can test this with the indiemelon site, and the following codes:

- Variation:
_qr-ok-variation-blue-medium


- UPC-E:
_barcode_upce_0123469

**Comment by toupper** [comment]
Thanks a lot for the review @iamgabrielma !

Could you please create an issue with that information? It seems important, I will take care of it as part of milestone 1 👍

**Comment by iamgabrielma** [comment]
Sure, logged here: https://github.com/woocommerce/woocommerce-ios/issues/9913

**Pull request merged**